src/rewrite.rs

```diff
@@ -256,46 +281,102 @@ where
 {
     /// Apply many substititions.
     ///
-    /// This method should call [`apply_one`] for each match and then
-    /// unify the results with the matched eclass.
-    /// This should return a list of [`Id`]s where the union actually
-    /// did something.
+    /// This method should call [`apply_one`] for each match,
+    /// then call [`union_results`] to combine the results.
+    ///
+    /// It returns the ids resulting from the calls to [`union_results`].
     ///
     /// The default implementation does this and should suffice for
     /// most use cases.
     ///
     /// [`apply_one`]: Applier::apply_one()
-    fn apply_matches(&self, egraph: &mut EGraph<L, N>, matches: &[SearchMatches]) -> Vec<Id> {
+    fn apply_matches(
+        &self,
+        egraph: &mut EGraph<L, N>,
+        searcher: &Arc<dyn Searcher<L, N> + Sync + Send>,
+        matches: &[SearchMatches],
+        rule_name: &str,
+    ) -> Vec<Id> {
         let mut added = vec![];
         for mat in matches {
             for subst in &mat.substs {
-                let ids = self
-                    .apply_one(egraph, mat.eclass, subst)
-                    .into_iter()
-                    .filter_map(|id| {
-                        let (to, did_something) = egraph.union(id, mat.eclass);
-                        if did_something {
-                            Some(to)
-                        } else {
-                            None
-                        }
-                    });
-                added.extend(ids)
+                let ids = self.apply_one(egraph, mat.eclass, subst);
+                let searcher_ast = searcher.get_ast(egraph, mat.eclass, subst);
+                let unions = self.union_results(
+                    egraph,
+                    mat.eclass,
+                    ids,
+                    searcher_ast.as_ref(),
+                    subst,
+                    rule_name,
+                );
+                added.extend(unions)
             }
         }
         added
     }
 
+    /// For patterns, get the ast directly as a reference.
+    fn get_pattern_ast(&self) -> Option<&PatternAst<L>> {
+        None
+    }
+
+    /// Return the pattern corresponding to the applier.
+    ///
+    /// This pattern is commonly used for proof-production mode.
+    fn get_ast(
+        &self,
+        egraph: &mut EGraph<L, N>,
+        eclass: Id,
+        subst: &Subst,
+    ) -> Option<PatternAst<L>> {
+        None
+    }
+
+    /// Unions the eclasses after a match has been applied.
+    ///
+    /// This should return a list of [`Id`]s of eclasses which
+    /// had unions applied to them.
+    /// [`union_results`]: Applier::union_results()
+    fn union_results(
```

Comment:
`apply_matches` by default now calls union_results instead of unioning directly, giving users a hook to do the right thing in proof production mode